### PR TITLE
feat(s3): support generating endpoints from multi-region access point

### PR DESCRIPTION
**Pull request by AllanZhengYP**
### Issue
Reverts: cfb3fff268eb393ece2ea50956a6f8610ec51c40.

This change marks `@aws-sdk/signature-v4-crt` package is peer dependency. It makes sure the SDK won't install the package out-of-box. Because the `@aws-sdk/signature-v4-crt` package contains native code, it takes up 10MB on disk, and it is only required for signing MRAP request. So for users requiring this feature, they need to install the package explicitly.

TODO: Create `Known Issues` to the README.md too document users to install `@aws-sdk/signature-v4-crt` package explicitly

### Testing
Manual test(Lambda, Lambda with bundler, Browser with Webpack), Unit test

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: clients/client-s3/models/models_1.ts [hunk 1]
```diff
@@ -514,6 +514,7 @@ export interface RestoreObjectRequest {
...
    *          <p>When using this action with Amazon S3 on Outposts, you must direct requests to the S3 on Outposts hostname. The S3 on Outposts hostname takes the form <i>AccessPointName</i>-<i>AccountId</i>.<i>outpostID</i>.s3-outposts.<i>Region</i>.amazonaws.com. When using this action using S3 on Outposts through the Amazon Web Services SDKs, you provide the Outposts bucket ARN in place of the bucket name. For more information about S3 on Outposts ARNs, see <a href="https://docs.aws.amazon.com/AmazonS3/latest/userguide/S3onOutposts.html">Using S3 on Outposts</a> in the <i>Amazon S3 User Guide</i>.</p>
+   * <p>Note: To supply the Multi-region Access Point(MRAP) to Bucket, you need to install the "aws-crt" package sparately. For more information, please go to https://github.com/aws/aws-sdk-js-v3#known-issues</p>
```
Comment (AllanZhengYP) [inline clients/client-s3/models/models_1.ts:517]:
```suggestion
   * <p>Note: To supply the Multi-region Access Point (MRAP) to Bucket, you need to install the "aws-crt" package separately. 
   * For more information, please go to https://github.com/aws/aws-sdk-js-v3#known-issues</p>
```

File: packages/middleware-sdk-s3/src/S3SignatureV4.ts [hunk 2]
```diff
@@ -23,30 +23,54 @@ export type S3SignerV4Init = SignatureV4Init &
  */
 export class S3SignatureV4 implements RequestPresigner, RequestSigner {
   private readonly sigv4Signer: SignatureV4;
-  private readonly sigv4aSigner?: CrtSignerV4;
+  private sigv4aSigner?: RequestSigner & RequestPresigner;
```
Comment (AllanZhengYP) [inline packages/middleware-sdk-s3/src/S3SignatureV4.ts:26]:
```suggestion
  private sigv4aSigner?: CrtSignerV4;
```

File: packages/middleware-sdk-s3/src/S3SignatureV4.ts [hunk 3]
```diff
@@ -23,30 +23,54 @@ export type S3SignerV4Init = SignatureV4Init &
...
-      if (!this.sigv4aSigner)
+      if (this.signerOptions.runtime !== "node")
         throw new Error("This request requires signing with SigV4Asymmetric algorithm. It's only available in Node.js");
+      if (!this.sigv4aSigner) {
+        const CrtSignerV4 = await expectSigv4aSigner();
+        this.sigv4aSigner = new CrtSignerV4({
+          ...this.signerOptions,
+          signingAlgorithm: 1,
+        });
+      }
```
Comment (AllanZhengYP) [inline packages/middleware-sdk-s3/src/S3SignatureV4.ts:44]: To remove duplicate code, move all them into `expectSigV4aSigner()`.

File: packages/middleware-sdk-s3/src/S3SignatureV4.ts [hunk 4]
```diff
@@ -23,30 +23,54 @@ export type S3SignerV4Init = SignatureV4Init &
...
     return this.sigv4Signer.presign(originalRequest, options);
   }
 }
+
+const expectSigv4aSigner = async (): Promise<new (options: CrtSignerV4Init & SignatureV4CryptoInit) => CrtSignerV4> => {
+  try {
+    return (await import("@aws-sdk/signature-v4-crt")).CrtSignerV4;
+  } catch (e) {
+    e.message =
+      `${e.message}\nPlease check if you have installed "@aws-sdk/signature-v4-crt" package explicitly. ` +
```
Comment (AllanZhengYP) [inline packages/middleware-sdk-s3/src/S3SignatureV4.ts:72]:
```suggestion
      `${e.message}\nPlease check if you have installed "@aws-sdk/signature-v4-crt" package explicitly. \n` +
```

File: packages/middleware-user-agent/src/user-agent-middleware.ts [hunk 5]
```diff
@@ -31,38 +31,39 @@ export const userAgentMiddleware =
...
-  ): BuildHandler<any, any> =>
-  async (args: BuildHandlerArguments<any>): Promise<BuildHandlerOutput<Output>> => {
-    const { request } = args;
-    if (!HttpRequest.isInstance(request)) return next(args);
-    const { headers } = request;
-    const userAgent = context?.userAgent?.map(escapeUserAgent) || [];
-    const defaultUserAgent = (await options.defaultUserAgentProvider()).map(escapeUserAgent);
-    const customUserAgent = options?.customUserAgent?.map(escapeUserAgent) || [];
+  ): BuildHandler<any, any> => {
+    return async (args: BuildHandlerArguments<any>): Promise<BuildHandlerOutput<Output>> => {
```
Comment (AllanZhengYP) [inline packages/middleware-user-agent/src/user-agent-middleware.ts:35]: Remove return block

File: packages/util-user-agent-node/src/is-crt-available.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    // (WebPack, Rollup) will try to bundle this optional dependency and fail to build if not exist.
+    // Thus this user agent key will only available in Node.js runtime.
+    if (typeof require === "function" && typeof module !== "undefined" && module.require && require("aws-crt")) {
+      // Validate `module` to make sure this is not in a `require.js` scope.
+      // TODO: load package version.
+      return ["md/crt-avail"];
+    }
+    return null;
+  } catch (e) {
+    console.log("||||||", process.versions, e);
```
Comment (AllanZhengYP) [inline packages/util-user-agent-node/src/is-crt-available.ts:16]:
```suggestion
```

File: packages/util-user-agent-node/src/is-crt-available.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import { UserAgentPair } from "@aws-sdk/types";
```
Comment (AllanZhengYP) [inline packages/util-user-agent-node/src/is-crt-available.ts:1]: TODO: add unit test

**Review by trivikr — APPROVED** [review]

**Pull request merged**